### PR TITLE
Incorporating Ngen workflow into CFE

### PR DESCRIPTION
**Pull request by stcui007**
This PR adds a workflow that run Ngen test within CFE with a realization configuration file the same as performing a ngen run within the framework with geojson data.

## Additions

The workflow file:
.github/workflows/ngen_integration.yaml

Realization configuration file:
realizations/realization_cfe_pet_ngenCI.json

## Removals

-

## Testing

github tests
-
## Notes

-

## Todos

-

## Checklist

- [x] PR has an informative and human-readable title
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [x] Code follows project standards (link if applicable)
- [x] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future todos are captured in comments
- [x] Visually tested in supported browsers and devices (see checklist below :point_down:)
- [x] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [x] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Target Environment support

- [ ] Windows
- [x] Linux
- [ ] Browser

### Accessibility

- [x] Keyboard friendly
- [x] Screen reader friendly

### Other

- [ ] Is useable without CSS
- [ ] Is useable without JS
- [ ] Flexible from small to large screens
- [ ] No linting errors or warnings
- [ ] JavaScript tests are passing

### CODE REVIEW

**Review by ajkhattak — APPROVED** [review]
my only concern was the modifications to the original `realization_cfe_pet.json` file, but now I see that you have reverted it. The other point, which is not that important but, I would suggest renaming your `ngen_realization_cfe_pet.json` to `realization_cfe_pet_ngenCI.json`  (or something like that) to avoid any confusion.

**Comment by PhilMiller** [comment]
Please fix the formatting of the PR description. In the future, for the checklists, it's not even necessary to 'check' them off through editing the text - you can just check them from the main PR Conversation page.

**Comment by PhilMiller** [comment]
The PR description says that `realizations/realization_cfe_pet.json` is modified. The previous PR did, but this one doesn't - it adds a new file instead. Please correct that.

**Comment by stcui007** [comment]
The latest commit ```821703d``` restore the ```realization_cfe_pet.json``` to its original form and added another realization config ```ngen_realization_cfe_pet.json``` for running workflow. The name may change, though, as one reviewer suggested. This is intended to not affect any native cfe test while adding the automated test.

**Review by PhilMiller — CHANGES_REQUESTED** [review]
Please correct the PR description to match the contents (and fix its formatting so it renders), or mark the PR as a draft until it's ready to be reviewed.

**Comment by madMatchstick** [comment]
@PhilMiller & @stcui007 - 

Just now looking at this workflow update.  Is the only outstanding item reforming the PR description format?  It would be very valuable to the team to have this pushed.  Thanks for your work on this!

File: .github/workflows/ngen_integration.yaml [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+      # Build Ngen and save some files
+      - name: Build Ngen
+        id: ngen_id1
+        uses: ./.github/actions/ngen-build
+        with:
+          targets: "ngen"
+          build-cores: ${{ env.LINUX_NUM_PROC_CORES }}
+          #is this required for this test?
+          bmi_c: 'ON'
+          bmi_fortran: 'OFF'
```
Comment (PhilMiller) [inline .github/workflows/ngen_integration.yaml:69]: Setting `bmi_fortran: 'OFF'` here seems to be inconsistent with building `extern/iso_c_fortran_bmi` above. Is it using Fortran support or not?

File: .github/workflows/ngen_integration.yaml [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+      matrix:
+        os: [ubuntu-latest, macos-latest]
+      fail-fast: false
+    runs-on: ${{ matrix.os }}
+
+    # Steps represent a sequence of tasks that will be executed as part of the job
+    steps:
+      # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
+      - name: checkout actions in another repo
+        uses: actions/checkout@v3
```
Comment (PhilMiller) [inline .github/workflows/ngen_integration.yaml:35]: An action further down is using `checkout@v4`. Is there a reason to use an older, inconsistent version here?

File: .github/workflows/ngen_integration.yaml [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+          # Move ngen build artifacts to temp directory
+          mv ${{ steps.ngen_id1.outputs.build-dir }} ${{runner.temp}}/ngen-build/
+          mv .github ${{runner.temp}}/.github
+          mv extern ${{runner.temp}}/extern
+          mv data ${{runner.temp}}/data
+
+      - name: Run surfacebmi plus cfebmi
+        run: |
+          echo "end of ngen_integration testing"
+
```
Comment (PhilMiller) [inline .github/workflows/ngen_integration.yaml:83]: This step doesn't actually do anything. 
```suggestion
```

File: .github/workflows/ngen_integration.yaml [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+          mv ${{runner.temp}}/ngen-build ./ngen-build
+          mv ${{runner.temp}}/extern/sloth extern
+          cp -r ${{runner.temp}}/extern/evapotranspiration extern
+          mv ${{runner.temp}}/data data
+
+          # Run ngen with CFE with PET
+          inputfile='realizations/ngen_realization_cfe_pet.json'
+          ./ngen-build/ngen ./data/catchment_data.geojson "cat-27" ./data/nexus_data.geojson "nex-26" $inputfile
+
+      # The following remove a post processing error
```
Comment (PhilMiller) [inline .github/workflows/ngen_integration.yaml:103]: This looks non-sensical. What's the error, and why/how does re-cloning the ngen repo fix it?

**Comment by PhilMiller** [comment]
The PR description says it's adding `realizations/ngen_realization_cfe_pet.json`, but the change actually adds `realization_cfe_pet_ngenCI.json`

File: .github/workflows/ngen_integration.yaml [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+          make -C cmake_build
+
+          # Move ngen build and other files to appropriate directory to run ngen
+          mv ${{runner.temp}}/ngen-build ./ngen-build
+          mv ${{runner.temp}}/extern/sloth extern
+          cp -r ${{runner.temp}}/extern/evapotranspiration extern
+          mv ${{runner.temp}}/data data
+
+          # Run ngen with CFE with PET
+          inputfile='realizations/ngen_realization_cfe_pet.json'
```
Comment (PhilMiller) [inline .github/workflows/ngen_integration.yaml:100]:
```suggestion
          inputfile='realizations/ngen_realization_cfe_pet_ngenCI.json'
```
It seems that this should match the file that's actually added

Comment (stcui007) [reply .github/workflows/ngen_integration.yaml]: That's probably why the test failed.

**Review by PhilMiller — APPROVED** [review]

**Comment by stcui007** [comment]
I think it should run when a PR is created and on subsequent push of new
commit. It is supposed to work on CFE repo. Hopefully, it will work in a
similar way the ngen repo does.

**Pull request merged**